src/hww.c

```diff
@@ -149,28 +168,62 @@ static size_t _api_info(uint8_t* buf)
     return current - buf;
 }
 
+static void _clear_state(void)
+{
+    _state.blocking_op = HWW_BLOCKING_OP_NONE;
+    _state.unlock_wf = NULL;
+}
+
+static void _lock(hww_blocking_op_t op)
+{
+    _state.blocking_op = op;
+    usb_processing_lock(usb_processing_hww());
+}
+
+/**
+ * Clears the current state and unlocks the USB stack (if locked).
+ */
+static void _unlock(void)
+{
+    _clear_state();
+    if (_state.blocking_op != HWW_BLOCKING_OP_NONE) {
+        usb_processing_unlock();
+    }
+}
+
+static void _unlock_finished_cb(bool result, void* param)
+{
+    (void)param;
+    _state.unlock_result = result;
+    _state.unlock_wf = NULL;
+}
+
 /**
  * Executes the HWW packet.
  * @param[in] in_packet The incoming HWW packet.
  * @param[in] out_packet The outgoing HWW packet.
  * @param[in] max_out_len The maximum number of bytes that the outgoing HWW packet can hold.
+ * @return true if the response has been completed, false otherwise.
  */
-static void _process_packet(const in_buffer_t* in_req, buffer_t* out_rsp)
+static void _process_packet(const in_buffer_t* in_req, hww_packet_rsp_t* out_rsp)
 {
+    out_rsp->status = HWW_RSP_ACK;
     if (in_req->len >= 1) {
         switch (in_req->data[0]) {
         case OP_ATTESTATION:
-            _api_attestation(in_req, out_rsp);
+            _api_attestation(in_req, &out_rsp->buffer);
             return;
         case OP_UNLOCK:
             if (!memory_is_initialized()) {
-                out_rsp->data[0] = OP_STATUS_FAILURE_UNINITIALIZED;
+                out_rsp->buffer.data[0] = OP_STATUS_FAILURE_UNINITIALIZED;
+                out_rsp->buffer.len = 1;
+                return;
             } else {
-                out_rsp->data[0] =
-                    workflow_unlock_blocking() ? OP_STATUS_SUCCESS : OP_STATUS_FAILURE;
+                _state.unlock_wf = workflow_unlock(&_unlock_finished_cb, NULL);
```

Comment:
workflow_unlock finishes immediately if the device is already unlocked. With the current retry logic, the unlock request is delayed by whatever retry-interval the client has, which is noticeable.

How to solve this?

Reply:
I would just add a check in `hww.c` that avoids calling `workflow_unlock` entirely if the device is already unlocked. The only other case in which the workflow will return immediately is the path with `!memory_is_initialized()`, which is rare enough (IMHO) to make a 200ms irrelevant. (Also, if `OP_UNLOCK` is only sent during pairing, I don't think 200ms would make a difference there as well, but that's again IMHO).

Reply:
Fixed in e9a5715